src/mat/impls/inv/matinv.c

```diff
@@ -50,14 +50,14 @@ static PetscErrorCode MatInvGetRegularizationType_Inv(Mat imat,MatRegularization
   PetscFunctionReturn(0);
 }
 
+//TODO MOVE THIS TO permonmatnullspace.c
+//TODO Passing the PC is awkward. We will want to avoid that but it needs deeper refactoring.
 #undef __FUNCT__
-#define __FUNCT__ "MatInvComputeNullSpace_Inv"
+#define __FUNCT__ "MatComputeNullSpaceMat"
 #if defined(PETSC_HAVE_MUMPS)
-static PetscErrorCode MatInvComputeNullSpace_Inv(Mat imat)
+PetscErrorCode MatComputeNullSpaceMat(Mat K, PC pc, MatOrthType orthType, MatOrthForm orthForm, Mat *R_new)
```

Comment:
We should not return R_New. Instead we should set it to K.

Reply:
Like this it is slightly more general and less awkward for some use cases. I also like the clear separation of input and output. I don't see the point of making K an inout argument and changing its properties under the hood.

Reply:
> Like this it is slightly more general and less awkward for some use cases. 

What use case?

I think it is weird to compute the nullspace, which you argue that is the property of a matrix, and not set it to the matrix.

Reply:
I think that a function BlaComputeBla should just compute what is desired and return that without any unseen side effects.

Consider `QPTDualize()`. I think what I have here,
```
TRY( MatComputeNullSpaceMat(K,pc,MAT_ORTH_GS,MAT_ORTH_FORM_EXPLICIT,&R) );
TRY( MatSetNullSpaceMat(K,R) );
TRY( MaDestroy(&R) );
```
is clearer and less error-prone than
```
TRY( MatComputeNullSpaceMat(K,pc,MAT_ORTH_GS,MAT_ORTH_FORM_EXPLICIT) );
//Unclear that R is updated internally.
//Old R needs to be destroyed so the pointer can be dangling now!

TRY( MatGetNullSpaceMat(K,&R) );
```
and requires less documentation.

I think a single clearly defined [mutator](https://en.wikipedia.org/wiki/Mutator_method) function is in general a preferred practice from various reasons. For example, one could hang some callback to the Set function but forget about the other functions altering the property.

Reply:
There exist a good use case. The R returned by MUMPS is dense so in a memory constrained situation I may not want to keep it in K and destroy it as soon as I do not need it.